lib/Streaming/ByteString.hs

```diff
@@ -410,15 +423,44 @@ testNull p@(Chunk bs rest) = if B.null bs
 
 -- | Remove empty ByteStrings from a stream of bytestrings.
 denull :: Monad m => Stream (ByteStream m) m r -> Stream (ByteStream m) m r
-denull = hoist (run . maps effects) . separate . mapped nulls
-{-# INLINE denull #-}
+{-# INLINABLE denull #-}
+denull = loop . Right
+  where
+    -- Scan each substream, dropping empty chunks along the way.  As soon as a
+    -- non-empty chunk is found, just apply the loop to the next substream in
+    -- the terminal value via fmap.  If Empty comes up before that happens,
+    -- continue the current stream instead with its denulled tail.
+    --
+    -- This implementation is tail recursive:
+    -- * Recursion via 'loop . Left' continues scanning an inner ByteStream.
+    -- * Recursion via 'loop . Right' moves to the next substream.
+    --
+    -- The old version below was shorter, but noticeably slower, especially
+    -- when empty substreams are frequent:
```

Comment:
How did you measure the perf improvement?

Reply:
I ran two tests:
- One with a large file with no empty lines, where `lines` produces a stream with no *nulls*. In this case runtime for the old `denull` algorithm was approximately 1.5x the new runtime.
- A second with a large file where 90% of the lines were empty, so that `denull` eliminates most lines. Here the runtime for the old `denull` was around 4-5x the runtime of the new.

The code I was timing was (new):
```haskell
module Main (main) where
import qualified Streaming.ByteString as Q
import qualified Streaming.ByteString.Char8 as Q8

main :: IO ()
main = Q.stdout
    $ Q.intercalate (Q.singleton 0x0a)
    $ Q.denull
    $ Q8.lines
    $ Q.stdin
```
vs. (old):
```haskell
module Main (main) where
import qualified Streaming.ByteString as Q
import qualified Streaming.ByteString.Char8 as Q8
import Streaming.ByteString.Internal (ByteStream(..))
import Streaming.Internal (Stream(..))
import qualified Streaming.Prelude as S
import qualified Streaming as S

denull :: Monad m => Stream (ByteStream m) m r -> Stream (ByteStream m) m r
denull = S.hoist (S.run . S.maps Q.effects) . S.separate . S.mapped Q.nulls

main :: IO ()
main = Q.stdout
    $ Q.intercalate (Q.singleton 0x0a)
    $ denull
    $ Q8.lines
    $ Q.stdin
```

Reply:
Considering how proactive you've been with benchmarking, it might make sense to make an official `bench` component for this project. Thoughts?

Reply:
> Considering how proactive you've been with benchmarking, it might make sense to make an official `bench` component for this project. Thoughts?

That would be cool, but I probably won't have the cycles to do it justice...